### PR TITLE
Allow registering custom `ColorResolver`s

### PR DESCRIPTION
**Pull request by diesieben07**
`ColorResolver` is used to create biome-based block coloring. The `BlockAndTintGetter`  given to `BlockColor` cannot directly be used to obtain the biome, it must be done through `ColorResolver`, which will allow the game to implement caching and also apply the selected biome blend. Any ColorResolver instance must be registered for this cache though. This PR adds an event to allow mods to register their own.

### CODE REVIEW

File: src/main/java/net/minecraftforge/client/ForgeHooksClient.java [hunk 1]
```diff
@@ -318,6 +321,23 @@ public static void onItemColorsInit(ItemColors itemColors, BlockColors blockColo
         ModLoader.get().postEvent(new RegisterColorHandlersEvent.Item(itemColors, blockColors));
     }
 
+    private static ImmutableList<ColorResolver> colorResolvers;
```
Comment (marchermans) [inline src/main/java/net/minecraftforge/client/ForgeHooksClient.java]: Please move this into its own management class.

**Review by marchermans — CHANGES_REQUESTED** [review]

**Review by marchermans — APPROVED** [review]

File: src/main/java/net/minecraftforge/client/ColorResolverManager.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+        colorResolvers = builder.build();
+    }
+
+    /**
+     * Register a {@link BlockTintCache} for every registered {@link ColorResolver} into the given target map.
+     *
+     * @param level the level to use
+     * @param target the map to populate
+     */
+    public static void registerBlockTintCaches(ClientLevel level, Map<ColorResolver, BlockTintCache> target)
```
Comment (sciwhiz12) [inline src/main/java/net/minecraftforge/client/ColorResolverManager.java:40]: Should this be marked as `@ApiStatus.Internal`? Is there a reason for a mod to call this method?

File: src/test/java/net/minecraftforge/debug/client/CustomColorResolverTest.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+import net.minecraftforge.eventbus.api.IEventBus;
+import net.minecraftforge.eventbus.api.SubscribeEvent;
+import net.minecraftforge.fml.common.Mod;
+import net.minecraftforge.fml.javafmlmod.FMLJavaModLoadingContext;
+import net.minecraftforge.registries.DeferredRegister;
+import net.minecraftforge.registries.ForgeRegistries;
+import net.minecraftforge.registries.RegistryObject;
+
+@Mod(CustomColorResolverTest.MOD_ID)
+public class CustomColorResolverTest
```
Comment (sciwhiz12) [inline src/test/java/net/minecraftforge/debug/client/CustomColorResolverTest.java:31]: Please provide a javadoc for this class which explains what feature this tests, how to test it, and what the expected results are from said test.

This should also be added to the test sourceset's `mods.toml` for now.

**Review by sciwhiz12 — CHANGES_REQUESTED** [review]

File: src/main/java/net/minecraftforge/client/ColorResolverManager.java (hunk 2 above)
Comment (diesieben07) [reply src/main/java/net/minecraftforge/client/ColorResolverManager.java:40]: My reason for not marking it as internal was for mods that do custom rendering (maybe even something like Optifine) needing to call it.

**Comment by diesieben07** [comment]
Test mod without biome blend

And with biome blend

**Review by sciwhiz12 — APPROVED** [review]

**Comment by diesieben07** [comment]
Updated to 1.19.2

**Review by sciwhiz12 — APPROVED** [review]

**Pull request merged**